### PR TITLE
Implemented `SetExponent` function

### PR DESCRIPTION
**Pull request by parth121101**
fixes: #3608

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2012,6 +2013,83 @@ namespace Fraction {
     }
 }  // namespace Fraction
 
+namespace Set_Exponent {
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2016]:
```suggestion
namespace SetExponent {
```

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2012,6 +2013,83 @@ namespace Fraction {
...
+    static ASR::expr_t *eval_Set_Exponent(Allocator &al, const Location &loc,
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2017]:
```suggestion
    static ASR::expr_t *eval_SetExponent(Allocator &al, const Location &loc,
```

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2012,6 +2013,83 @@ namespace Fraction {
...
+                exponent = ((ix >> 52) & 0x7ff) - 1022;
+                double result1 =  x * std::pow((2), (-1*(exponent)));
+                double result = result1 * std::pow((2), I);
+                return make_ConstantWithType(make_RealConstant_t, result, arg_type, loc);
+            }
+        }
+        return nullptr;
+    }
+
+    static inline ASR::expr_t* instantiate_Set_Exponent(Allocator &al, const Location &loc,
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2062]:
```suggestion
    static inline ASR::expr_t* instantiate_SetExponent(Allocator &al, const Location &loc,
```

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -2012,6 +2013,83 @@ namespace Fraction {
...
+        arg_types_exp.push_back(al, arg_types[0]);
+
+        Vec<ASR::call_arg_t> new_args_exp; new_args_exp.reserve(al, 1);
+        ASR::call_arg_t arg1; arg1.loc = loc; arg1.m_value = args[0];
+        new_args_exp.push_back(al, arg1);
+
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[1], ASR::cast_kindType::IntegerToReal, return_type, nullptr));
+
+        ASR::expr_t* func_call_fraction = Fraction::instantiate_Fraction(al, loc, scope, arg_types_exp, return_type, new_args_exp, 0);
+        body.push_back(al, b.Assignment(result, r_tMul(func_call_fraction, rPow(i2r(i(2, int32),return_type),cast, return_type), return_type)));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:2085]:
```suggestion

        ASR::expr_t* func_call_fraction = Fraction::instantiate_Fraction(al, loc, scope, arg_types_exp, return_type, new_args_exp, 0);
        body.push_back(al, b.Assignment(result, r_tMul(func_call_fraction, rPow(i2r(i32(2),return_type),i2r(args[1], return_type), return_type), return_type)));
```

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 5]
```diff
@@ -396,6 +396,12 @@
             "ret_type_arg_idx": 0
         },
     ],
+    "Set_Exponent": [
```
Comment (Pranavchiku) [inline src/libasr/intrinsic_func_registry_util_gen.py:399]:
```suggestion
    "SetExponent": [
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 6]
```diff
@@ -61,6 +61,7 @@ inline std::string get_intrinsic_name(int x) {
         INTRINSIC_NAME_CASE(Ble)
         INTRINSIC_NAME_CASE(Exponent)
         INTRINSIC_NAME_CASE(Fraction)
+        INTRINSIC_NAME_CASE(Set_Exponent)
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:64]:
```suggestion
        INTRINSIC_NAME_CASE(SetExponent)
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 7]
```diff
@@ -230,6 +231,8 @@ namespace IntrinsicElementalFunctionRegistry {
             {&Exponent::instantiate_Exponent, &Exponent::verify_args}},
         {static_cast<int64_t>(IntrinsicElementalFunctions::Fraction),
             {&Fraction::instantiate_Fraction, &Fraction::verify_args}},
+        {static_cast<int64_t>(IntrinsicElementalFunctions::Set_Exponent),
+            {&Set_Exponent::instantiate_Set_Exponent, &Set_Exponent::verify_args}},
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:235]:
```suggestion
        {static_cast<int64_t>(IntrinsicElementalFunctions::SetExponent),
            {&SetExponent::instantiate_SetExponent, &SetExponent::verify_args}},
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 8]
```diff
@@ -477,6 +480,8 @@ namespace IntrinsicElementalFunctionRegistry {
             "exponent"},
         {static_cast<int64_t>(IntrinsicElementalFunctions::Fraction),
             "fraction"},
+        {static_cast<int64_t>(IntrinsicElementalFunctions::Set_Exponent),
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:483]:
```suggestion
        {static_cast<int64_t>(IntrinsicElementalFunctions::SetExponent),
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 9]
```diff
@@ -676,6 +681,7 @@ namespace IntrinsicElementalFunctionRegistry {
                 {"ble", {&Ble::create_Ble, &Ble::eval_Ble}},
                 {"exponent", {&Exponent::create_Exponent, &Exponent::eval_Exponent}},
                 {"fraction", {&Fraction::create_Fraction, &Fraction::eval_Fraction}},
+                {"set_exponent", {&Set_Exponent::create_Set_Exponent, &Set_Exponent::eval_Set_Exponent}},
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:684]:
```suggestion
                {"set_exponent", {&SetExponent::create_SetExponent, &SetExponent::eval_SetExponent}},
```

File: src/libasr/pass/intrinsic_functions.h [hunk 10]
```diff
@@ -61,6 +61,7 @@ enum class IntrinsicElementalFunctions : int64_t {
     Ble,
     Exponent,
     Fraction,
+    Set_Exponent,
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_functions.h:64]:
```suggestion
    SetExponent,
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
We follow camel case in name of intrinsic functions. Also, requested a simpler way to implement it. You may follow various macros and avoid `cast` expr.

File: src/libasr/pass/intrinsic_functions.h [hunk 11]
```diff
@@ -2012,6 +2013,81 @@ namespace Fraction {
...
+            float x = ASR::down_cast<ASR::RealConstant_t>(args[0])->m_r;
+            int32_t I = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+            int32_t exponent;
+            if (x == 0.0) {
+                exponent = 0;
+                float result1 =  x * std::pow((2), (-1*(exponent)));
+                float result = result1 * std::pow((2), I);
+                return make_ConstantWithType(make_RealConstant_t, result, arg_type, loc);
+            }
+            else{
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:2031]: Format things this way:
```suggestion
            } else {
```

File: src/libasr/pass/intrinsic_functions.h [hunk 12]
```diff
@@ -2012,6 +2013,81 @@ namespace Fraction {
...
+            double x = ASR::down_cast<ASR::RealConstant_t>(args[0])->m_r;
+            int64_t I = ASR::down_cast<ASR::IntegerConstant_t>(args[1])->m_n;
+            int64_t exponent;
+            if (x == 0.0) {
+                exponent = 0;
+                double result1 =  x * std::pow((2), (-1*(exponent)));
+                double result = result1 * std::pow((2), I);
+                return make_ConstantWithType(make_RealConstant_t, result, arg_type, loc);
+            }
+            else{
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:2050]:
```suggestion
            } else {
```

**Review by certik — APPROVED** [review]

**Pull request merged**